src/coreclr/jit/morph.cpp

```diff
@@ -12903,6 +12903,49 @@ GenTree* Compiler::fgOptimizeAddition(GenTreeOp* add)
 
             return add;
         }
+
+        // Fold (~x + 1) to -x.
+        if (op1->OperIs(GT_NOT) && op2->IsIntegralConst(1))
+        {
+            op1->SetOper(GT_NEG);
+            op1->SetVNsFromNode(add);
+            DEBUG_DESTROY_NODE(op2);
+            DEBUG_DESTROY_NODE(add);
+            return op1;
+        }
+
+        // Fold (~x + GT_CNS_INT/GT_CNS_LNG) to ((GT_CNS_INT/GT_CNS_LNG - 1) - x).
+        // Ex. (~x + 2) = (1 - x)
+        if (op1->OperIs(GT_NOT) && op2->IsIntegralConst() &&
+            (op2->AsIntCon()->IconValue() > 1 || op2->AsIntCon()->IconValue() < 1))
+        {
+        
+            add->SetOper(GT_SUB);
+        
+            add->gtOp1              = op2;
+            GenTreeIntCon* constOp2 = op2->AsIntCon();
+            constOp2->SetValueTruncating(constOp2->IconValue() - 1);
+        
+            add->gtOp2 = op1->AsOp()->gtGetOp1();
+            DEBUG_DESTROY_NODE(op1);
+        
+            return add;
+        }
+
+        // Fold (-1 - (~x)) to x
+        if ((op1->OperIs(GT_NEG) && op1->AsOp()->gtGetOp1()->OperIs(GT_NOT)) &&
+            op2->IsIntegralConst(-1)) 
```

Comment:
I think it would be better to do this as two separate transforms `-(~x) => x + 1` and `~(-x) => x - 1`. However, I think you should check if there are occurrences in SPMI to see if we should bother with it.

Reply:
I had played with it and tried different variations, but there were not any occurrences in SPMI, so I reverted the changes.